packages/core/src/anchor/processing-loop.ts

```diff
@@ -0,0 +1,74 @@
+export class Deferred<T = void> {
+  readonly resolve: (t: T) => void
+  readonly reject: (e: Error) => void
+  readonly then: Promise<T>['then']
+  readonly catch: Promise<T>['catch']
+
+  constructor() {
+    let dResolve: Deferred<T>['resolve']
+    let dReject: Deferred<T>['reject']
+
+    const promise = new Promise<T>((resolve, reject) => {
+      dResolve = resolve
+      dReject = reject
+    })
+    this.resolve = dResolve
+    this.reject = dReject
+    this.then = promise.then.bind(promise)
+    this.catch = promise.catch.bind(promise)
+  }
+}
+
+export class ProcessingLoop<T> {
+  private readonly source: AsyncGenerator<T>
+  private readonly handleValue: (value: T) => Promise<void>
+  #processing: Promise<void> | undefined
+  #deferred: Deferred
+  #abortController: AbortController
+
+  constructor(source: ProcessingLoop<T>['source'], onValue: ProcessingLoop<T>['handleValue']) {
+    this.source = source
+    this.handleValue = onValue
+    this.#processing = undefined
+    this.#deferred = new Deferred()
+    this.#abortController = new AbortController()
+  }
+
+  start() {
```

Comment:
What are the cases where this is not just started right away? Looking through the code, it seems like this should just always be called in the constructor.

Reply:
...unless we explicitly disable it (enableLoop = false) to avoid interference with some tests. Also, it is not quite compatible with the way other stateful parts of Ceramic work, like storage.